situp-plugins/mapdb-processor-state/src/main/java/com/amazon/situp/plugins/processor/state/MapDbProcessorState.java

```diff
@@ -0,0 +1,118 @@
+                        .treeMap(dbName)
+                        .counterEnable() //Treemap doesnt keep size counter by default
+                        .keySerializer(SIGNED_BYTE_ARRAY_SERIALIZER)
+                        .valueSerializer(Serializer.JAVA).create();
+    }
+
+
+    @Override
+    public void put(byte[] key, V value) {
+        map.put(key, value);
+    }
+
+    public void putAll(final Map<byte[], V> data) {
+        map.putAll(data);
+    }
+
+    @Override
+    public V get(byte[] key) {
+        return map.get(key);
+    }
+
+    @Override
+    public Map<byte[], V> getAll() {
+        return map;
+    }
+
+    @Override
+    public <R> List<R> iterate(BiFunction<byte[], V, R> fn) {
+        final List<R> returnList = new ArrayList<>();
+        map.entryIterator().forEachRemaining(
+                entry -> returnList.add(fn.apply(entry.getKey(), entry.getValue()))
+        );
+        return returnList;
+    }
+
+    public <R> List<R> iterate(BiFunction<byte[], V, R> fn, final int segments, final int index) {
+        if(map.size() == 0) {
+            return Collections.EMPTY_LIST;
+        }
+        final byte[][]iterationEndpoints = getIterationEndpoints(segments, index);
+        final List<R> returnList = new ArrayList<>();
+        map.entryIterator(iterationEndpoints[0], true, iterationEndpoints[1], false).forEachRemaining(
+                entry -> returnList.add(fn.apply(entry.getKey(), entry.getValue()))
+        );
+        return returnList;
+    }
+
+    /**
+     * Gets iteration endpoints by taking the lowest and highest key and splitting the keyrange into segments.
+     * These endpoints are an approximation of segments, and segments are guaranteed to cover the entire key range,
+     * but there is no guarantee that all segments contain an equal number of elements.
+     * @param segments Number of segments
+     * @param index Index to find segment endpoints for
+     * @return byte[][] containing the two endpoints
+     */
+    private byte[][] getIterationEndpoints(final int segments, final int index) {
+        final BigInteger lowEnd = new BigInteger(map.firstKey());
+        final BigInteger highEnd = new BigInteger(map.lastKey());
+        final BigInteger step = highEnd.subtract(lowEnd).divide(new BigInteger(String.valueOf(segments)));
+        final byte[] lowIndex = lowEnd.add(step.multiply(new BigInteger(String.valueOf(index)))).toByteArray();
+        final byte[] highIndex =
+                index == segments - 1 ?
+                        highEnd.add(new BigInteger("1")).toByteArray() :
+                        lowEnd.add(step.multiply(new BigInteger(String.valueOf(index+1)))).toByteArray();
+        return new byte[][]{lowIndex, highIndex};
+    }
+
+    @Override
+    public long size() {
+        return map.size();
+    }
+
+    @Override
+    public void clear() {
+        map.clear();
+    }
+
+    @Override
+    public void close() {
+        map.close();
+    }
```

Comment:
nit: From the first version of interface I have some odd feelings on this `clear` and `close`. I think we should just support one interface which is `delete`.  Delete works for any state implementation. Also, clear is expensive and close doesn't define cleaning up of resource. I agree mapdb support deleteOnClose, but other implementations won't.

Reply:
That makes sense. I'll merge the interfaces into delete()